crates/worker/src/partition/state_machine/command_interpreter/mod.rs

```diff
@@ -1358,6 +1362,189 @@ where
+                        Entry::PeekPromise(PeekPromiseEntry { key, .. }) =
+                            journal_entry.deserialize_entry_ref::<Codec>()?
+                    );
+
+                    if let Some(service_id) =
+                        invocation_metadata.invocation_target.as_keyed_service_id()
+                    {
+                        // Load state and write completion
+                        let promise_metadata = state.get_promise(&service_id, &key).await?;
+
+                        let completion_result = match promise_metadata {
+                            Some(PromiseMetadata {
+                                state: PromiseState::Completed(result),
+                            }) => result.into(),
+                            _ => CompletionResult::Empty,
+                        };
+
+                        Codec::write_completion(&mut journal_entry, completion_result.clone())?;
+
+                        // Forward completion
+                        effects.forward_completion(
+                            invocation_id,
+                            Completion::new(entry_index, completion_result),
+                        );
+                    } else {
+                        warn!(
+                            "Trying to process entry {} for a target that has no promises",
+                            journal_entry.header().as_entry_type()
+                        );
+                        effects.forward_completion(
+                            invocation_id,
+                            Completion::new(entry_index, CompletionResult::Empty),
+                        );
+                    }
+                }
+            }
+            EnrichedEntryHeader::CompletePromise { is_completed, .. } => {
+                if !is_completed {
+                    let_assert!(
+                        Entry::CompletePromise(CompletePromiseEntry {
+                            key,
+                            completion,
+                            ..
+                        }) = journal_entry.deserialize_entry_ref::<Codec>()?
+                    );
+
+                    if let Some(service_id) =
+                        invocation_metadata.invocation_target.as_keyed_service_id()
+                    {
+                        // Load state and write completion
+                        let promise_metadata = state.get_promise(&service_id, &key).await?;
+
+                        let completion_result = match promise_metadata {
+                            None => {
+                                // Just register the promise completion
+                                effects.put_promise(
+                                    service_id,
+                                    key,
+                                    PromiseMetadata {
+                                        state: PromiseState::Completed(completion),
+                                    },
+                                );
```

Comment:
What is the lifetime of these promises? When are they gonna be cleaned up? Or are they kept around forever?

Reply:
see comment below.